[WIP] Create runner-installation.adoc (#4645)
circleci/circleci-docs: jekyll/_cci2/runner-installation.adoc

= Installing the CircleCI Runner
:page-layout: classic-docs
:page-liquid:
:icons: font
:toc: macro
:toc-title:

This page describes the steps required to install CircleCI Runner on Linux and macOS platforms.

=== Supported Platforms

At present, the supported platforms are:

> - Ubuntu 18.04 running on an x86_64 (AMD64) or ARM64 architecture

Reviewer: ```suggestion
- Ubuntu 18.04 running on an x86_64 (AMD64) or ARM64 architecture.
```

- macOS Catalina (10.15.6+) or above, running on an x86_64 architecture.

Running in a container is not supported at this time.

=== Prerequisites

==== Authentication

> In order to complete this process you'll need to create a namespace, resource-class and authentication token:

Reviewer 2: ```suggestion
In order to complete this process you will need to create a namespace, resource-class and authentication token:
```

> . Install the circleci command-line tool.

Reviewer 2: link to doc on installing the CLI?

Author: Added installation link :)

. Create a namespace for your organization's runner resources under:

Each organization can only create a single namespace.

> - If you already use orbs, this will be same namespace as the orbs use. Use the following command: `circleci namespace create <name> <vcs-type> <org-name>` (e.g. if your GitHub URL is `https://github.com/circleci`, then use: `circleci namespace create my-namespace github circleci`.

Reviewer: ```suggestion
- If you already use orbs, this will be the same namespace as the orbs use. Use the following command: `circleci namespace create <name> <vcs-type> <org-name>` (e.g. if your GitHub URL is `https://github.com/circleci`, then use: `circleci namespace create my-namespace github circleci`.
```

> - If you already use orbs, this will be same namespace as the orbs use. Use the following command: `circleci namespace create <name> <vcs-type> <org-name>` (e.g. if your GitHub URL is `https://github.com/circleci`, then use: `circleci namespace create my-namespace github circleci`.

Reviewer 2: Formatting seems a bit off here, is this all part of one step?

Author: Fixed formatting

> . Create a resource class for your runner in the above namespace:

Reviewer 2: For numbered steps to stay aligned correctly with subsequent paragraphs you need to use the following syntax:

```
. direction for numbered step
+
next paragraph content, still under the same number
. next numbered step
```

if that makes sense

Author: Fixed

This will print a generated Runner config including the authentication token. Note that the token cannot be retrieved again, so store it safely.

==== Installation Tooling

The installation process assumes that the following utilities are installed on the system:

- curl (installed by default on macOS)
- sha256sum (installed as part of coreutils on Linux apt/yum, macOS via brew)
- systemd version 235+ (Linux only)
- Installation also requires permissions to create a user, and create directories under /opt.

=== Job Running Requirements

Running jobs requires that the following tools are available on the machine:

- tar
- gzip
- coreutils (Linux only)
- git (recommended, but not required)

=== Installation

==== Download the Launch Agent Binary and Verify the Checksum

The Launch Agent can be installed using the following script in a bash shell.

This will use /opt/circleci as the base install location.

First, set one of these variables as appropriate for for your installation target.

===== For Linux x86_64

`platform=linux/amd64`

===== For Linux ARM64

`platform=linux/arm64`

===== For macOS x86_64

> `platform=darwin/amd64`

Reviewer 2: I would maybe put this in a table instead of under headings?

Author: Added a table. Thanks for the suggestion. This is so much better :)

And then run the following steps to download, verify and install the binary.

```sh
prefix=/opt/circleci
sudo mkdir -p "$prefix/workdir"
base_url="https://circleci-binary-releases.s3.amazonaws.com/circleci-launch-agent"
echo "Determining latest version of CircleCI Launch Agent"
agent_version=$(curl "$base_url/release.txt")
echo "Using CircleCI Launch Agent version $agent_version"
echo "Downloading and verifying CircleCI Launch Agent Binary"
curl -sSL "$base_url/$agent_version/checksums.txt" -o checksums.txt
file="$(grep -F "$platform" checksums.txt | cut -d ' ' -f 2)"
file="${file:1}"
mkdir -p "$platform"
echo "Downloading CircleCI Launch Agent: $file"
curl --compressed -L "$base_url/$agent_version/$file" -o "$file"
echo "Verifying CircleCI Launch Agent download"
sha256sum --check --ignore-missing checksums.txt && chmod +x "$file"; sudo cp "$file" "$prefix/circleci-launch-agent" || echo "Invalid checksum for CircleCI Launch Agent, please try download again"
```

=== Platform-specific instructions

Please refer to the platform-specific installation instructions:

> - linux

Reviewer: should these be linking to something?

Author: Added TOC so user could directly go to the desired installation section.

- macOS

=== Configuration file reference

A YAML file is used to configure the Launch Agent; how it communicates with our servers and how it will launch the Task Agent.

The configuration file uses the following format, the parameters are explained in more detail below:

```sh
api:
  auth_token: AUTH_TOKEN
runner:
  name: RUNNER_NAME
runner.name
```

> This is a unique name assigned to this particular running Launch Agent, we recommend using the hostname of the machine. It will be used to identify the agent when viewing statuses and job results in the CircleCI UI.

Reviewer: I'm not sure it's clear what "This" is referring to. Should this sentence start with `RUNNER_NAME` is a unique name...

Reviewer: or rather, perhaps this is missing the `==== runner.name` heading.

Author: fixed

==== api.auth_token

This is a token used to identify the Launch Agent to CircleCI, it will be provided by your Customer Success Manager. An existing token may be shared among many installations, but only allows a particular resource_class to be specified.

==== runner.command_prefix

> This allows you to customize how the Task Agent process is launched, our example uses the launch-task script provided below.

Reviewer: ```suggestion
This allows you to customize how the Task Agent process is launched; our example uses the launch-task script provided below.
```

Reviewer: (or `.`)

==== runner.working_directory

This allows you to control the default working directory used by each job. If the directory already exists, Task Agent will need permissions to write to it. If the directory does not exist, then the Task Agent will need permissions to create it. If `%s` is present in the value, it will be replaced with a different value for each job. These directories will not be automatically removed.

==== runner.cleanup_working_directory

This allows for control of working directory cleanup after each job. The default value is `false`.

==== runner.max_run_time

> This can be used to override the default maximum duration the Task Agent will run each job. The value is a string with the following unit identifiers `h`, `m` or `s` for hour minute and seconds respectively:

Reviewer: ```suggestion
This can be used to override the default maximum duration the Task Agent will run each job. The value is a string with the following unit identifiers `h`, `m` or `s` for hour minute and seconds respectively. Consider these valid examples:
```

Reviewer: If you commit this suggestion, you will also need to remove line 146 ( `A few valid examples`)

Author: Fixed

= Installing the CircleCI Runner
:page-layout: classic-docs
:page-liquid:
:icons: font
:toc: macro
:toc-title:

This page describes how to install CircleCI Runner on Linux and macOS platforms.

toc::[]

=== Prerequisites

==== Supported Platforms

Currently, the following platforms are supported with CircleCI Runner:

- Ubuntu 18.04 running on an x86_64 (AMD64) or ARM64 architecture.

> - macOS Catalina (10.15.6+) or above, running on an x86_64 architecture.

Reviewer 3: We need to clarify this — I have some internal documentation that we can use for it. Will message you.

Author: Fixed

Author: Updated this content based on your messaging.

> CAUTION: Running in a container is not currently supported.

Reviewer 3: We’ve added instructions to the preview docs for running it in a container.

Author: Added Docker installation content to page.

Author: Added to document

Author: Added to document

==== Authentication

In order to complete this process you will need to create a namespace, resource-class and authentication token by performing the steps listed below:

. Install the circleci command-line tool.

. Create a namespace for your organization's runner resources under:

NOTE: Each organization can only create a single namespace. If you already use orbs, this namespace will be the same namespace as the orbs use. Use the following command: `circleci namespace create <name> <vcs-type> <org-name>` (e.g. if your GitHub URL is `https://github.com/circleci`, then use: `circleci namespace create my-namespace github circleci`.

> . Create a resource class for your runner for your namespace using the following command: `circleci runner resource-class create <resource-class> <description>` (e.g. `circleci runner resource-class create my-namespace/my-resource-class my-description`)

Reviewer 3: Worth explaining what a resource class is here? Or linking to a Definitions section?

Author: fixed

= Installing the CircleCI Runner
:page-layout: classic-docs
:page-liquid:
:icons: font
:toc: macro
:toc-title:
version:
- Cloud

NOTE: CircleCI runner is currently only available if you are on the Scale Plan. Please reach out to your sales representative for information on how to sign up for the Scale Plan.

This page describes how to install CircleCI runner on Linux and macOS platforms.

toc::[]

> === Prerequisites

Reviewer 2: I think all the heading levels can go up one, so this one should be `==`

Author: fixed

Author: Fixed

==== Support Levels

> There are currently three support levels for CircleCI Runner:

Reviewer 2: ```suggestion
There are currently three support levels for CircleCI runner:
```

* Supported
* Untested

> * Untested

Reviewer 2: Unsupported?

Author: fixed

Reviewer 4: "Untested" is intentional - we may or may not support this via their DCE

===== Supported

* CircleCI has built and tested runner on a given platform.
* Support expectation:

> +

Reviewer 2: ```suggestion
```

> ** DCEs can provide support.

Reviewer 2: DCE is probably an internal term? Maybe we should use something more generic? - might be wrong on that though

Author: Fixed

===== Unsupported

* CircleCI will not provide a runner binary capable of running CircleCI runner on the platform.
* Support expectation:
+
** CircleCI runner will not work on the platform.
** A customer can try running CircleCI runner on the platform, but are unlikely to succeed.
** No CircleCI documentation or DCE support.

==== Supported Platforms

The following platforms have been validated to work with CircleCI runner:

* Intel + Ubuntu
* Arm64 + Ubuntu
* Intel + macOS

===== Untested Platforms

The following platforms have not been tested, but are believed to work with CircleCI runner:

* Intel + non-Ubuntu Linux (Debian, Red Hat, SUSE, etc)
* Arm64 + non-Ubuntu Linux (Debian, Red Hat, SUSE, etc)
* Docker

NOTE: CircleCI provides no documentation for these platforms.

===== Unsupported Platforms

The following platforms are not currently supported to work with CircleCI runner:

* Intel + Windows
* Arm + Windows
* Apple Silicon + macOS
* Arm32 + Linux
* Run natively in Kubernetes/Rancher/other container orchestration solutions

==== Authentication

> In order to complete this process you will need to create a namespace, https://circleci.com/docs/2.0/optimizations/#resource-class [resource-class] and authentication token by performing the steps listed below:

Reviewer 2: ```suggestion
In order to complete this process you will need to create a namespace, https://circleci.com/docs/2.0/optimizations/#resource-class[resource-class] and authentication token by performing the steps listed below:
```

Reviewer 4: I don't think the link to that resource class page is relevant in this context - it talks about optimisations

> 1. <<local-cli#installation,Install the circleci command-line tool>>.

Reviewer 2: @thomasr8 example of asciidoc link to another file in the same directory :-)

> 2. Create a namespace for your organization's runner resources under:

Reviewer 2: Feels like there's something missing at the end of this line?

Author: Fixed

Author: Fixed

4. Create a token for authenticating the above resource-class by using the following command: `circleci runner token create <resource-class> <nickname>` (e.g. `circleci runner token create my-namespace/my-resource-class my-token`). This will print a generated Runner configuration including the authentication token.

CAUTION: The token cannot be retrieved again, so be sure to store it safely.

==== Installation Tooling

The installation process assumes you have installed the following utilities on your system:

* curl (installed by default on macOS)
* sha256sum (installed as part of coreutils on Linux apt/yum, macOS via brew)
* systemd version 235+ (Linux only)
* permissions to create a user, and create directories under /opt.

=== Job Running Requirements

Running jobs requires you have the following tools available on your machine:

* tar
* gzip
* coreutils (Linux only)
* git (recommended, but not required)

=== Installation

==== Download the Launch Agent Binary and Verify the Checksum

The Launch Agent can be installed using the following script, which will use `opt/circleci` as the base install location.

> First, set one of these variables as appropriate for for your installation target.

Reviewer 2: ```suggestion
First, set one of these variables as appropriate for your installation target.
```
